torch/csrc/distributed/c10d/default_comm_hooks.cpp

```diff
@@ -46,7 +46,14 @@ c10::intrusive_ptr<c10::ivalue::Future> FP16CompressCommHook::runHook(
 c10::intrusive_ptr<c10::ivalue::Future> _AllReduceBySumCommHook::runHook(
     GradBucket& bucket) {
   std::vector<at::Tensor> tensors = {bucket.getBufferRef()};
+#ifdef IS_NCCL_EXP
+  // sparse gradient indices update case
+  AllreduceOptions opts = AllreduceOptions();
+  opts.sparseIndices = bucket.getGlobalUniqueId().value();
+  return state_->allreduce_sparse(tensors, opts)->getFuture();
```

Comment:
i might be confused, but does this mean we always call sparse allreduce if IS_NCCL_EXP = true (even if we have a dense tensor)? 

Have we tested the case where IS_NCCL_EXP=true and we're allreducing a dense tensor?

also I'm confused on this - if we have the dispatcher, why do we need to call `allreduce_sparse` explicitly here, shouldn't we call just `allreduce` and the dispatcher figures out which implementation to call?

Reply:
This is a good catch. I'm updating the code to only call into allreduce and utilize the dispatcher